src/core/symbology/qgssymbol.cpp

```diff
@@ -116,57 +119,69 @@ QPolygonF QgsSymbol::_getLineString( QgsRenderContext &context, const QgsCurve &
   }
 
   // remove non-finite points, e.g. infinite or NaN points caused by reprojecting errors
-  pts.erase( std::remove_if( pts.begin(), pts.end(),
-                             []( const QPointF point )
+  pts.filterVertices( []( const QgsPoint point )
   {
-    return !std::isfinite( point.x() ) || !std::isfinite( point.y() );
-  } ), pts.end() );
+    if ( point.is3D() )
+    {
+      return std::isfinite( point.x() ) && std::isfinite( point.y() ) && std::isfinite( point.z() );
+    }
+    else
+    {
+      return std::isfinite( point.x() ) && std::isfinite( point.y() );
+    }
+  } );
 
   if ( clipToExtent && nPoints > 1 && context.flags() & QgsRenderContext::ApplyClipAfterReprojection )
   {
     // early clipping was not possible, so we have to apply it here after transformation
     const QgsRectangle e = context.mapExtent();
     const double cw = e.width() / 10;
     const double ch = e.height() / 10;
-    const QgsRectangle clipRect( e.xMinimum() - cw, e.yMinimum() - ch, e.xMaximum() + cw, e.yMaximum() + ch );
-    pts = QgsClipper::clippedLine( pts, clipRect );
+    const QgsBox3d clipRect( QgsPoint( e.xMinimum() - cw, e.yMinimum() - ch ), QgsPoint( e.xMaximum() + cw, e.yMaximum() + ch ) ); // TODO also need to be clipped according to z axis
+    pts = QgsClipper::clipped3dLine( pts, clipRect );
   }
 
-  QPointF *ptr = pts.data();
-  for ( int i = 0; i < pts.size(); ++i, ++ptr )
+  QPolygonF out = pts.asQPolygonF();
+  QPointF *ptr = out.data();
+  for ( int i = 0; i < out.size(); ++i, ++ptr )
   {
     mtp.transformInPlace( ptr->rx(), ptr->ry() );
   }
 
-  return pts;
+  return out;
 }
 
 QPolygonF QgsSymbol::_getPolygonRing( QgsRenderContext &context, const QgsCurve &curve, const bool clipToExtent, const bool isExteriorRing, const bool correctRingOrientation )
 {
   const QgsCoordinateTransform ct = context.coordinateTransform();
   const QgsMapToPixel &mtp = context.mapToPixel();
 
-  QPolygonF poly = curve.asQPolygonF();
+  // clone...
+  QgsLineString poly;
+  QgsPointSequence seq;
+  curve.points( seq );
+  poly.setPoints( seq );
```

Comment:
As above -- we'll need two separate implementations of _getPolygonRing, one for the purely 2d case and one where z needs considering

Reply:
done!